Fix: sometimes, changes to subsections are not reflected in the Studio Search Index

bradenmacdonald opened the pull request:
## Description

There is a race condition that causes some changes to not be reflected in the Studio search index.

## Technical Details

What's happening is that the XBlock is updated with `modulestore.update_item()` in the initial request, that function [immediately triggers the `XBLOCK_UPDATED` signal](https://github.com/openedx/edx-platform/blob/73d3995dae42526844e729d30c846cfb7c1e3c0a/xmodule/modulestore/mixed.py#L824-L842). However, when our handler [processes that](https://github.com/openedx/edx-platform/blob/73d3995dae42526844e729d30c846cfb7c1e3c0a/openedx/core/djangoapps/content/search/api.py#L389-L415) [in a worker](https://github.com/openedx/edx-platform/blob/73d3995dae42526844e729d30c846cfb7c1e3c0a/openedx/core/djangoapps/content/search/handlers.py#L78-L81), even though there is no active MySQL transaction, the modulestore "bulk operation" _may_ still be active, so the actual modulestore data in mongo/mysql hasn't been updated yet. The data only [gets committed](https://github.com/openedx/edx-platform/blob/73d3995dae42526844e729d30c846cfb7c1e3c0a/xmodule/modulestore/split_mongo/split.py#L284-L291) [when the outermost bulk operation scope is exited](https://github.com/openedx/edx-platform/blob/73d3995dae42526844e729d30c846cfb7c1e3c0a/xmodule/modulestore/__init__.py#L286).

I couldn't find any existing mechanism to listen for that "actual commit" of _draft_ content, because the code seems to assume that other parts of the system only need to know when new content has been _published_, not when new draft content has been committed to the DB. So I had to add that as a new capability.

## Testing instructions

In `cms/envs/devstack.py` set `CELERY_ALWAYS_EAGER = False`.

1. Rename any subsection in a course
2. Open the course search (from the header of the course authoring MFE, while viewing the same course outline)
3. Search for the name of the subsection
4. In some cases (especially if it's not the first time you've made a change since the CMS and CMS worker were restarted), you will see that the name of the subsection in the search index is out of date.

We have focused testing on renaming subsections, but this likely affects other types of edits.

## Deadline

TBD

bradenmacdonald committed:
fix: don't call signal handlers like XBLOCK_UPDATED before commit

openedx-webhooks commented:
Thanks for the pull request, @bradenmacdonald! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

yusuf-musleh committed:
feat: Add remaining XBLOCK_CREATED event to on_commit

yusuf-musleh committed:
fix: Add checks before deleting bulk ops

yusuf-musleh commented on xmodule/modulestore/split_mongo/split.py:
```diff
@@ -233,11 +233,13 @@ def _clear_bulk_ops_record(self, course_key):
+                del self._active_bulk_ops.records[
+                    course_key.replace(org=None, course=None, run=None, branch=None)
+                ]
 
```
@bradenmacdonald The reason this [test case](https://github.com/openedx/edx-platform/blob/7d7a18d28bfc794b168ccb2c7b84e23034ebda7d/cms/djangoapps/contentstore/tests/test_contentstore.py#L1132-L1144) was failing was because for some reason the bulk op record is not there when attempting to clear it as part of [self._clear_bulk_ops_record](https://github.com/openedx/edx-platform/pull/34800/files#diff-ce330aeb39241bcb7cee96b41906443b85005e9f9acb84c02c0d1a751f8025aeR319) call. I didn't just want to add this check, I dug deeper to see what is actually causing the issue, especially since it used to work before without these checks.

Oddly enough, after narrowing it down to this [line](https://github.com/openedx/edx-platform/pull/34800/files#diff-ce330aeb39241bcb7cee96b41906443b85005e9f9acb84c02c0d1a751f8025aeR325), when it gets the bulk ops record it (which it does), but the error then occurs in the clearing step. If I remove this line, and just call  `fn()` right away it works fine. 

Before spending more time on this and digging more into it, I just wanted to make sure if my concerns are in the right place, i.e. there is a deeper issue and I should continue investigating, or is this "fix", i.e. adding these checks is a fine approach?

bradenmacdonald replied:
@yusuf-musleh Mmmm, that "fix" looks suspicious to me. I prefer not to change split modulestore behavior any more than is necessary, because it can have unforeseen consequences.

What I suspect may be the issue is actually in `BulkOperationsMixin`. If you look at it, most of the existing methods take a parameter `ignore_case=False`, but I didn't implement that for the new `on_commit_changes_to` method. So I suspect that when `create_child` or `update_item` is being called in this case, it has `ignore_case=True` in its kwargs, but we're not passing that into `on_commit_changes_to`, which results in that method creating a different bulk ops record. Or something like that. Otherwise, I'm not even sure how this change could possibly be affecting that test.

yusuf-musleh replied:
@bradenmacdonald Agreed, the issue is likely with the underlying implementation related to `ignore_case`. 

> Otherwise, I'm not even sure how this change could possibly be affecting that test.

Especially since, I tried changing the testcase from `course_number.upper() -> course_number.lower()`, just to see if the test passes (since the test checks that the course fails to be created) and it indeed passed.  That could explain why the test is failing.

However there doesn't seem to be a straightforward way to access or determine `ignore_case` inside the [update_item](https://github.com/openedx/edx-platform/pull/34800/files#diff-e2292d98eab04e5ce19b5c784091e24167aa697427772e83752c41f5310abad6R835) method so it can be passed in to the `on_commit_changed_to`.

So I've been toying with the idea of possibly adding a new `ignore_case` field to the `BulkOperationsMixin` that could then be accessible from `update_item` method, it's not quite working yet as I'm figuring out the best way to initialize it/set the value without not make too many changes. Could that be a valid approach, or we'd rather not such changes here, as it might be over complication?

bradenmacdonald replied:
@yusuf-musleh The bug turns out to be a pretty subtle one related to the use of `defaultdict` and the fact that calling `_get_bulk_ops_record` _creates_ a new record. So the code in `on_commit_changes_to` had an unintentional side-effect. You can fix it with this change, without needing to modify any other parts of split modulestore:

```diff
         """
         Call some callback when the currently active bulk operation has saved
         """
-        bulk_ops_record = self._get_bulk_ops_record(course_key)
-        if bulk_ops_record.active:
+        # Check if a bulk op is active. If so, defer fn(); otherwise call it immediately.
+        # Note: calling _get_bulk_ops_record() here and then checking .active can have side-effects in some cases
+        # because it creates an entry in the defaultdict if none exists, so we check if the record is active using
+        # the same code as _clear_bulk_ops_record(), which doesn't modify the defaultdict.
+        # so we check it this way:
+        if course_key and course_key.for_branch(None) in self._active_bulk_ops.records:
+            bulk_ops_record = self._active_bulk_ops.records[course_key.for_branch(None)]
             bulk_ops_record.defer_until_commit(fn)
```

yusuf-musleh replied:
@bradenmacdonald That was indeed it! Thanks for all your help on this. Overall this bug has been quite the challenge to figure out, especially with all the subtleties 😅

yusuf-musleh committed:
fix: Change check for active bulk op on_commit

Initially we used `_get_bulk_ops_record()` however that had side-effects of creating records since it uses a defaultdict internally. So we swapped the implementation similar how we check in `_clear_bulk_ops_record()`

bradenmacdonald commented:
@ormsbee Would you be able to review this PR? We'd also like to backport it to Redwood, as it's a bugfix.

ChrisChV reviewed (approved):
@yusuf-musleh Looks good :+1:
- [X] I tested this: I followed the testing instructions with: Sections, Units and components
- [X] I read through the code and considered the security, stability and performance implications of the changes.
- [X] I tested that the UI can be used with a keyboard only (tab order, keyboard controls).
- [X] Includes tests for bugfixes and/or features added.
- [X] Includes documentation

ormsbee reviewed (approved):
Nice! This is honestly a lot cleaner than I thought it would be when you described the problem. 😛

bradenmacdonald commented:
Haha yeah @ormsbee I'm happy with how the fix shaped up; we were able to implement it without needing to change too much.

openedx-webhooks commented:
@bradenmacdonald 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/34800&entry.1484655318=https://github.com/bradenmacdonald&entry.752974735=2024-05-14+19:41&entry.1917517419=2024-05-22+17:29&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
